### PR TITLE
[SofaImplicitOdeSolver] Rewrite of the static Newton-Raphson ODE solver

### LINKED ISSUE #224: Share in SOFA the StaticSolver
A clean version implementing a correct incremental loading and a correct Newton loop.

### PR DESCRIPTION
**Pull request by jnbrunet**
This commit rewrites the static ODE solver, implements two additional convergence criteria, improves the documentation of every step found in the "solve" method, and adds useful doxygen documentation of the Newton algorithms executed by the component.

Up until this commit, the only convergence criteria of the static ODE solver were absolute force residual norms, and absolute (displacement) increment norm.  Using absolute values for convergence criteria means that the value set by the user is dependent on the simulation scale. For example, setting an absolute displacement norm of 1e-5 might seem enough for a liver mesh set in millimetres, but is definitely not adequate for the same liver using this time a mesh scaled in metres. This is a source of user error, and is probably why many FEM software will use relative convergence criteria.

The two additional criteria are:

1. The relative residual norm |R|/|R0| where |R| is the norm of the force vector after the last Newton iteration, and |R0| is the norm of the force vector after the first Newton iteration.
2. The relative displacement norm |dx|/|U| where |dx| is the norm of the latest solution vector, and |U| is the norm of the sum over all previous solution vectors of the current time step.

To remove any confusion between absolute and relative criteria, the convergence data field "correction_tolerance_threshold" and "residual_tolerance_threshold" were renamed for "absolute_correction_tolerance_threshold" and "absolute_residual_tolerance_threshold", respectively. Note that to keep a compatibility layer with older scene files, these old data field names are still available, but a deprecated warning is shown to the user, inviting him to update its scene.

[ci-depends-on https://github.com/sofa-framework/SofaPython3/pull/147]

EDIT: Fix #224

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by jnbrunet** [comment]
Setting back to WIP after a quick review
1. absolute and relative residual are inverted in the convergence checks
2. Missing the divergence test when the residual is growing
3. I will add unit tests for relative norms

**Comment by jnbrunet** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaImplicitOdeSolver/src/SofaImplicitOdeSolver/StaticSolver.cpp [hunk 1]
```diff
@@ -62,145 +76,356 @@ void StaticSolver::parse(sofa::core::objectmodel::BaseObjectDescription* arg)
...
+
+        // Part VI. Stop timers and print step information.
+        {
+            auto iteration_time = duration_cast<nanoseconds>(steady_clock::now() - t).count();
+
+            if (print_log) {
+                info << "Newton iteration #" << std::left << std::setw(5) << n_it
+                     << std::scientific
+                     << "  |R| = " << std::setw(12) << std::sqrt(R_squared_norm)
+                     << "  |R|/|R0| = " << std::setw(12) << std::sqrt(R_squared_norm / R0_squared_norm)
```
Comment (alxbilger) [inline SofaKernel/modules/SofaImplicitOdeSolver/src/SofaImplicitOdeSolver/StaticSolver.cpp]: Possible division by zero: `R0_squared_norm` could be zero

File: SofaKernel/modules/SofaImplicitOdeSolver/src/SofaImplicitOdeSolver/StaticSolver.cpp [hunk 2]
```diff
@@ -62,145 +76,356 @@ void StaticSolver::parse(sofa::core::objectmodel::BaseObjectDescription* arg)
...
+                     << "  |du| = " << std::setw(12) << std::sqrt(dx_squared_norm)
+                     << "  |du| / |U| = " << std::setw(12) << std::sqrt(dx_squared_norm / U_squared_norm)
```
Comment (alxbilger) [inline SofaKernel/modules/SofaImplicitOdeSolver/src/SofaImplicitOdeSolver/StaticSolver.cpp]: Possible division by zero: `U_squared_norm` could be zero
You check if `U_squared_norm < epsilon*epsilon` in the next Part.

**Review by alxbilger — COMMENTED** [review]
Excellent PR. Thanks a lot for all the details you wrote in the comment. It helps a lot to understand all the steps of a solver, and how to execute them in Sofa.

I am really concerned about [this operator](https://github.com/sofa-framework/sofa/blob/15cccb1eacacc31f9223ff9369e8240d1002d425/SofaKernel/modules/SofaCore/src/sofa/core/behavior/MultiMatrix.h#L71). It hides a lot of steps, and it is easy to miss it. So thanks again to highlight it in the comments.

Cosmetic: use Allman style (brace associated with a control statement on the next line)

**Comment by jnbrunet** [comment]
Thanks a lot for the review @alxbilger 

> I am really concerned about this operator. It hides a lot of steps, and it is easy to miss it.

Totally agree with you here.

**Comment by jnbrunet** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
Seems OK, could you just rebase/merge with master to remove the failed CUDA/whatever scenes? Just to be in the clear.

**Comment by jnbrunet** [comment]
[ci-build][with-all-tests]

**Comment by jnbrunet** [comment]
[ci-build]

**Comment by guparan** [comment]
[ci-build]

**Pull request merged**